### PR TITLE
[NTGDI] Check if face->charmap is not zero before accessing its encoding

### PR DESCRIPTION
**Pull request by julcar**
## Purpose

Attempting to start DevExpress Ribbon Notepad makes crash ReactOS due to null pointer access, after some investigation, I found that `(face->charmap)` might return 0 sometimes, and we are not making this check.

This allows DevExpress Ribbon Notepad sample to start.

JIRA issue: [CORE-18091](https://jira.reactos.org/browse/CORE-18091) [CORE-18558](https://jira.reactos.org/browse/CORE-18558)

### CODE REVIEW

**Comment by learn-more** [comment]
There are a few places where it's checked for NULL already,
but all these places also assign a charmap in that case.
Sadly, there are some changes between how they manage that, but I suspect that they should all do the same.
(And that it should happen in all cases before `->charmap` is accessed.

https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=grep&sr=1&s=charmap+%3D%3D+NULL

**Comment by vicmarcal** [comment]
> There are a few places where it's checked for NULL already, but all these places also assign a charmap in that case. Sadly, there are some changes between how they manage that, but I suspect that they should all do the same. (And that it should happen in all cases before `->charmap` is accessed.
> 
> https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=grep&sr=1&s=charmap+%3D%3D+NULL

Mmm comparing all the uses, seems the first result in that link is following a different pattern (encoding checking) to assign a charmap  compared with the other 2 (which are blindly assigning the first one found). Any reasoning behind that? Just curious.

**Review by katahiromz — APPROVED** [review]

**Comment by KRosUser** [comment]
Does it fixes the refered Jira ticket ? Or is it a side fix ?

**Comment by julcar** [comment]
> Does it fixes the refered Jira ticket ? Or is it a side fix ?

It is a real fix, the program affected started to run after applying the fix

**Comment by KRosUser** [comment]
So good to merge since July, right ?

**Comment by KRosUser2** [comment]
@julcar : Still OK to go with this PR ?

**Review by learn-more — APPROVED** [review]

**Review by KRosUser2 — APPROVED** [review]

**Comment by KRosUser2** [comment]
Please note it also fixes https://jira.reactos.org/browse/CORE-18558

File: win32ss/gdi/ntgdi/freetype.c [hunk 1]
```diff
@@ -4501,6 +4501,8 @@ ftGetFontUnicodeRanges(PFONTGDI Font, PGLYPHSET glyphset)
     DWORD num_ranges = 0;
     FT_Face face = Font->SharedFace->Face;
 
+    if (!(face->charmap)) return 0;
```
Comment (tkreuzer) [inline win32ss/gdi/ntgdi/freetype.c:4504]:
```suggestion
    if (face->charmap == NULL)
    {
        DPRINT1("FIXME: No charmap selected! This is a BUG!\n");
        return 0;
    }
```

**Review by tkreuzer — CHANGES_REQUESTED** [review]
This is at best a band-aid. There is a lot wrong in the code. 

- With this fix ftGetFontUnicodeRanges would simply fail, instead of selecting the proper charmap.
- In some places the code alls FT_Set_Charmap, first acquiring, then instantly releasing the freetype lock. This is wrong. Either we set the charset only once and keep it immutable in the rest of the code, or all code that relies on it must run with the freetype lock help.
- Code duplication.

I'm ok with a quick-fix, but put at least a fat DPRINT1 there.

**Review by JoachimHenze — APPROVED** [review]

**Review by tkreuzer — APPROVED** [review]

**Pull request merged**